python/ml4ir/base/model/architectures/auto_dag_network.py

```diff
@@ -0,0 +1,355 @@
+from __future__ import annotations
+
+from typing import List, Dict, Union, Type
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from ml4ir.base.config.keys import FeatureTypeKey
+from ml4ir.base.features.feature_config import FeatureConfig
+from ml4ir.base.io.file_io import FileIO
+
+
+def get_layer_subclasses() -> Dict[str, Type[layers.Layer]]:
+    """Get mapping of {subclass-name: subclass} for all derivative classes of keras.layers.Layer"""
+
+    def full_class_name(cls):
+        """Get {package_name}.{class_name}"""
+        module = cls.__module__
+        if module == "builtins":
+            return cls.__qualname__  # avoid outputs like 'builtins.str'
+        return f"{module}.{cls.__qualname__}"
+
+    def get_sub_classes(cls):
+        """DFS traversal to get all subclasses of `cls` class"""
+        for subcls in cls.__subclasses__():
+            if subcls not in subclasses:
+                # Handles duplicates
+                subclasses[full_class_name(subcls)] = subcls
```

Comment:
@arvindsrikantan Updating a global variable in a recursive loop is probably not the best coding practice, but it should be fine here as this is not really reusable code.

Reply:
Agreed. Passing by reference is an option, but not necessary unless we want to reuse

Reply:
It's not global will be my excuse :P